README.md

```diff
@@ -1,19 +1,122 @@
+
+And then the gradient of `foo` is
+```Haskell
+grad_foo :: HasDelta r => Domain0 r -> (Domain0 r, r) 
+grad_foo ds = 
+   let ((result, _, _, _), value) = 
+         dReverse 1 fooM (ds, V.empty, V.empty, V.empty) 
+   in (result, value) 
+```
+
+As can be verified by computing the gradient at `(1.1,2.2,3.3)`:
+```Haskell
+>>> grad_foo (mkDual [1.1, 2.2, 3.3])
+[7.75, 1.23, 5.52]
+```
+
+## Computing Jacobians.
+Now let's consider a function from 'R^n -> R^m'.  We don't want the gradient, but instead the Jacobian.
+```Haskell
+-- A function that goes from `R^3` to `R^2`. 
+foo :: RealFloat a => (a,a,a) -> (a,a)
+foo (x,y,z) = 
+  let w = x * sin y 
+  in (atan2 z w, z * w)
+```
+TODO: show how the 2x3 Jacobian emerges from here
+
+# Forall shapes and sizes
+
+An additional feature of this library is a type system for tensor shape arithmetic.  Look at this beauty.
+```Haskell
+convMnistTwoS x                                 -- input images, shape (batch_size, c_in, h, w)
+              ker1 bias1                        -- layer1 kernel, shape (c_out, c_in, kh+1, kw+1); and bias, shape (c_out)
+              ker2 bias2                        -- layer2 kernel, shape (c_out, c_out, kh+1, kw+1); and bias, shape (c_out)
+              weigthsDense                      -- dense weights shape (n_hidden, c_out, ((h+kh)/2 + kh)/2, ((w+kw)/2 + kw)/2) 
+              biasesDense                       -- dense biases shape (n_hidden)
+              weigthsReadout biasesReadout      -- readout layer weights (10, n_hidden) and bias (10)
+              =                                 --> output (10, batch_size)
+  do
+  t1 <- convMnistLayerS ker1 (constant x) bias1
+  t2 <- convMnistLayerS ker2 t1 bias2
+  let m1 = mapS reshapeS t2
+      m2 = from2S (transpose2 (fromS2 m1))  -- TODO: add permuation transposeS
+      denseLayer = weigthsDense <>$ m2 + asColumnS biasesDense
+  denseRelu <- reluAct denseLayer
+  returnLet $ weigthsReadout <>$ denseRelu + asColumnS biasesReadout
+```
+But we don't just want the shapes in comments, we want them in the type system:
+```Haskell
+convMnistTwoS
+  :: forall kh kw num_hidden c_out h w c_in batch_size d r m.
+     ( KnownNat kh, KnownNat kw, KnownNat num_hidden, KnownNat c_out, KnownNat h, KnownNat w, KnownNat batch_size
+     , c_in ~ 1 -- greyscale
+     , 1 <= kh  -- kh greater than 1
+     , 1 <= kw  -- kw greater than 1
+     , DualMonad d r m )
+  => OS.Array '[batch_size, c_in, h, w] r
+  -> DualNumber d (OS.Array '[ c_out, c_in , kh + 1, kw + 1 ] r)
+  -> DualNumber d (OS.Array '[c_out] r)
+  -> DualNumber d (OS.Array '[ c_out, c_out, kh + 1, kw + 1 ] r)
+  -> DualNumber d (OS.Array '[c_out] r)
+  -> DualNumber d (OS.Array '[ num_hidden, c_out,
+                                 GHC.TypeLits.* ((h + kh) `Div` 2 + kh) `Div` 2
+                                 GHC.TypeLits.* ((w + kw) `Div` 2 + kw) `Div` 2 -- [buglet? kh->kw? -awf]
```

Comment:
Yes, this is a bug and it was not caught by GHC due to the not implemented `transpose2` TODO above (now it's caught). This beautifully undermines the whole story of correctness of shapes guaranteed by dependent types.

The flop was possible, because Haskell is not a dependently typed language, as Agda or Idris, but a wide-spectrum language, where you can mix dependent and normal typing. Unfortunately, it's not easy to detect if such mixing is taking place or not. The new `transpose2S` function is fully dependently typed, while the `from2S` used before is, in fact, a coercion of matrices to rank 2 tensors, so its very purpose is to override dependent typing.

In the code you cited, after the TODO is fixed, the only remaining lax function is `reshapeS` that, at most, could guarantee on the type level that the number of elements does not change after the tensor reshaping. However what the resulting ranks are is unknown and so the function is a coercion and can hide bugs. We could instead use function `flattenS1S` that would require that the resulting tensor is of rank 1 and so no flexibility would be left in the output of the function. That's probably worth doing just for the sake of this example. However, at some point we need to decide how many differentiable tensor manipulation operations we want to maintain and if it's not better, e.g.,  to constrain the use of `reshapeS` with `:: DualNumber d (OS.Array '[... * ... * ...])` instead of adding `flattenS1S` that does the very same thing but in one word.